File: tasks/translation/output/Qwen/QwQ-32B-AWQ/codenet/direct/inline/comment/direct/Java/Python/s685734377_commented.py
import sys

def main():
    n = int(sys.stdin.readline())
    a = []
    for _ in range(n):
        line = sys.stdin.readline().strip()
        a.append(list(line))
    
    P = 37
    powP = [1] * (n + 1)
    for i in range(1, n + 1):
        powP[i] = powP[i - 1] * P
    
    hashRow = [[0] * n for _ in range(n)]
    hashCol = [[0] * n for _ in range(n)]
    
    for i in range(n):
        for j in range(n):
            c_val = ord(a[i][j])
            for k in range(n):
                term = c_val * powP[(j - k + n) % n]
                hashRow[i][k] += term
                term_col = c_val * powP[(i - k + n) % n]
                hashCol[j][k] += term_col
    
    ans = 0
    for A in range(n):
        for B in range(n):
            ok = True
            for i in range(n):
                row_idx = (i + A) % n
                col_idx = (i + B) % n
                val_row = hashRow[row_idx][B]
                val_col = hashCol[col_idx][A]
                if val_row != val_col:
                    ok = False
                    break
            if ok:
                ans += 1
    print(ans)

if __name__ == '__main__':
    main()
# End of Code

File: tasks/translation/output/Qwen/QwQ-32B-AWQ/codenet/direct/inline/comment/direct/Java/Go/s685734377_commented.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

func main() {
	in := bufio.NewScanner(os.Stdin)
	var n int
	if !in.Scan() {
		fmt.Fprintln(os.Stderr, "Error reading input")
		os.Exit(1)
	}
	n, _ = strconv.Atoi(in.Text())

	grid := make([][]rune, n)
	for i := 0; i < n; i++ {
		if !in.Scan() {
			fmt.Fprintln(os.Stderr, "Error reading grid line")
			os.Exit(1)
		}
		grid[i] = []rune(in.Text())
	}

	P := int64(37)
	Q := int64(47)
	powP := make([]int64, n+1)
	powQ := make([]int64, n+1)
	powP[0] = 1
	powQ[0] = 1
	for i := 1; i <= n; i++ {
		powP[i] = powP[i-1] * P
		powQ[i] = powQ[i-1] * Q
	}

	hashRow := make([][]int64, n)
	hashCol := make([][]int64, n)
	for i := range hashRow {
		hashRow[i] = make([]int64, n)
	}
	for j := range hashCol {
		hashCol[j] = make([]int64, n)
	}

	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			for k := 0; k < n; k++ {
				exponent := (j - k + n) % n
				hashRow[i][k] += int64(grid[i][j]) * powP[exponent]
				exponent2 := (i - k + n) % n
				hashCol[j][k] += int64(grid[i][j]) * powQ[exponent2]
			}
		}
	}

	ans := 0
	for A := 0; A < n; A++ {
		for B := 0; B < n; B++ {
			ok := true
			for i := 0; i < n; i++ {
				row := (i + A) % n
				col := (i + B) % n
				if hashRow[row][B] != hashCol[col][A] {
					ok = false
					break
				}
			}
			if ok {
				ans++
			}
		}
	}
	fmt.Println(ans)
}

// End of Code
